### PR TITLE
Som plugin

### PR DESCRIPTION
**Pull request by LuisSanchez25**
_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_

## What does the code in this PR do / what does it improve?

Creates a new peaklet classification plugin that uses SOM to refine our current peaklet classification algorithm.

## Can you briefly describe how it works?

We first classify the data using out traditional peaklet classification algorithm, then all the data that gets classified as either an S1 or an S2 are passed to the SOM where the new data can be reclassified as type 0, 1 or 2, corresponding to the typical designation of junk, S1 and S2.

We also create a new context xenonnt_som which uses this pluggin instead of the usual peaklet classification plugin.

## Can you give a minimal working example (or illustrate with a figure)?

_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [x] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### _Notes on testing_
 - _Until the automated tests pass, please mark the PR as a draft._
 - _On the XENONnT fork we test with database access, on private forks there is no database access for security considerations._

All _italic_ comments can be removed from this template.

### CODE REVIEW

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+    #rechunk_on_save = immutabledict(
+    #    peaklet_classification_som=True,
+    #    som_peaklet_data=True)
+
+    depends_on = ('peaklets', 'peaklet_classification')
+
+    provides = ('peaklet_classification', 'som_peaklet_data')
+    #provides = ('peaklet_classification_som')
+    data_kind = {k: k for k in provides}
```
Comment (jmosbacher) [inline straxen/plugins/peaklets/peaklet_classification_som.py:38]: this makes no sense. Why are you creating new data kinds? arent you just adding columns to  peaklets?

File: straxen/plugins/merged_s2s/merged_s2s.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ class MergedS2s(strax.OverlapWindowPlugin):
     """
     __version__ = '1.0.2'
 
-    depends_on = ('peaklets', 'peaklet_classification', 'lone_hits')
+    depends_on = ('peaklets', 'peaklet_classification_som', 'lone_hits')
```
Comment (jmosbacher) [inline straxen/plugins/merged_s2s/merged_s2s.py:20]: I thought we said your plugin would be a drop-in replacement for the peaklet classification. So why are you adding it as an extra data_type?

File: straxen/contexts.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import typing as ty
 from pandas.util._decorators import deprecate_kwarg
 import socket
+from straxen.plugins.peaklets.peaklet_classification_som import PeakletClassificationSOM
```
Comment (jmosbacher) [inline straxen/contexts.py:11]: Please conform with the existing coding style for plugin management.
plugins are not imported explicitly.

File: straxen/contexts.py [hunk 4]
```diff
@@ -111,6 +112,18 @@ def xenonnt(cmt_version='global_ONLINE', xedocs_version=None,
...
 
 
+def xenonnt_som(cmt_version='global_ONLINE', xedocs_version=None,
+                _from_cutax=False, **kwargs):
+    """XENONnT context for the SOM"""
+
+    st = straxen.contexts.xenonnt(cmt_version=cmt_version, xedocs_version=xedocs_version,
+                                  _from_cutax=_from_cutax, **kwargs)
+    del st._plugin_class_registry['peaklet_classification']
+    st.register(PeakletClassificationSOM)
```
Comment (jmosbacher) [inline straxen/contexts.py:122]:
```suggestion
    st.register(straxen.PeakletClassificationSOM)
```

File: straxen/plugins/peaklets/__init__.py [hunk 5]
```diff
@@ -3,3 +3,6 @@
 
 from . import peaklet_classification
 from .peaklet_classification import *
+
+#from . import peaklet_classification_som
```
Comment (jmosbacher) [inline straxen/plugins/peaklets/__init__.py:7]: Please remove comments so the plugin is available in the straxen namespace

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
+import numpy as np
+import numpy.lib.recfunctions as rfn
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply straxen/plugins/peaklets/peaklet_classification_som.py:2]: please use 
```python
from numpy.lib import recfunctions as rfn
```

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+from scipy.spatial.distance import cdist
+from straxen.plugins.peaklets.peaklet_classification import PeakletClassification
```
Comment (jmosbacher) [inline straxen/plugins/peaklets/peaklet_classification_som.py:4]: remove this import

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+import numba
+
+import strax
+import straxen
+
+export, __all__ = strax.exporter()
+
+
+@export
+class PeakletClassificationSOM(PeakletClassification):
```
Comment (jmosbacher) [inline straxen/plugins/peaklets/peaklet_classification_som.py:14]:
```suggestion
class PeakletClassificationSOM(straxen.PeakletClassification):
```

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+    provides = ('peaklet_classification_som', 'som_peaklet_data')
+    data_kind = {k: k for k in provides}
+
+    # parallel = True
+
+    som_files = straxen.URLConfig(default='resource:///stor2/data/LS_data/SOM_data/som_data_v4.1.npz?fmt=npy')
+
+    def infer_dtype(self):
+        dtype = dict()
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply straxen/plugins/peaklets/peaklet_classification_som.py:43]:
```suggestion
        dtype = {}
```

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+                                      self.som_s0_array)
+        result['time'] = peaklets['time']
+        result['length'] = peaklets['length']
+        result['dt'] = peaklets['dt']
+        result['type'][mask_non_zero] = strax_type
+        #result['som_type'][mask_non_zero] = som_type + 1
+        return dict(peaklet_classification_som=result, som_peaklet_data=som_info)
+
+
+def recall_populations(dataset, weight_cube, SOM_cls_img, norm_factors):
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply straxen/plugins/peaklets/peaklet_classification_som.py:130]: please fix this

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+    Master function that should let the user provide a weightcube,
+    a reference img as a np.array, a dataset and a set of normalization factors.
+    In theory, if these 5 things are provided, this function should output
+    the original data back with one added field with the name "SOM_type"
+    weight_cube:      SOM weight cube (3D array)
+    SOM_cls_img:      SOM reference image as a numpy array
+    dataset:          Data to preform the recall on (Should be peaklet level data)
+    normfactos:       A set of 11 numbers to normalize the data so we can preform a recall
+    """
+    [SOM_xdim, SOM_ydim, SOM_zdim] = weight_cube.shape
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply straxen/plugins/peaklets/peaklet_classification_som.py:143]:
```suggestion
    xdim, ydim, zdim = weight_cube.shape
```

[the remark this answers is not in the document]
Comment (jmosbacher) [reply straxen/plugins/peaklets/peaklet_classification_som.py:143]: please remove all these CamelCase variable names. Use snake_case for variable names.

**Review by jmosbacher — CHANGES_REQUESTED** [review]
You are committing to **collaborative** project. Please conform to existing coding styles so that others can easily read your work.

File: straxen/plugins/peaklets/peaklet_classification_som.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    ref_map = np.zeros((xdim, ydim))
+    for color in np.arange(len(unique_colors)):
+        mask = np.all(np.equal(color_image, unique_colors[color, :]), axis=2)
+        indices = np.argwhere(mask)  # generates a 2d mask
+        for loc in np.arange(len(indices)):
+            ref_map[indices[loc][0], indices[loc][1]] = color
+    return ref_map
+
+
+def SOM_cls_recall(array_to_fill, data_in_som_fmt, weight_cube, reference_map):
```
Comment (jmosbacher) [inline straxen/plugins/peaklets/peaklet_classification_som.py:182]:
```suggestion
def som_cls_recall(array_to_fill, data_in_som_fmt, weight_cube, reference_map):
```

**Comment by LuisSanchez25** [comment]
The problem with changing the structure with how you suggested (just importing straxen and using straxen.PeakletClassification for the plugin) is that it creates a circular import error since straxen cant initialize it. This is also part of the reason I removed it from the __init__.py file since it can just be initialized later by the context, the other reason is that in [contexts](https://github.com/XENONnT/straxen/blob/master/straxen/contexts.py) if you look at line 88, it will register all plugins and since the SOM pluggin also provides peaklet classification I think it will cause problems. I could explicitly state not to fetch that config in the context where its registering everything but I think we probably do not want to hardcode that.

**Review by jmosbacher — APPROVED** [review]

**Comment by LuisSanchez25** [comment]
@jmosbacher all test have passed so I think this is ready to merge now

**Review by jmosbacher — APPROVED** [review]

**Comment by LuisSanchez25** [comment]
I cant merge the pull request so if you could merge it that would be great!

**Pull request merged**